tests/SpecTests/CrudSpecTest.php

```diff
@@ -0,0 +1,162 @@
+<?php
+
+namespace MongoDB\Tests\SpecTests;
+
+use stdClass;
+
+/**
+ * Transactions spec tests.
+ *
+ * @see https://github.com/mongodb/specifications/tree/master/source/crid
+ */
+class CrudSpecTest extends FunctionalTestCase
+{
+    /* These should all pass before the driver can be considered compatible with
+     * MongoDB 4.2. */
+    private static $incompleteTests = [
+        'aggregate-merge: Aggregate with $merge' => 'PHPLIB-438',
+        'aggregate-merge: Aggregate with $merge and batch size of 0' => 'PHPLIB-438',
+        'aggregate-merge: Aggregate with $merge and majority readConcern' => 'PHPLIB-438',
+        'aggregate-merge: Aggregate with $merge and local readConcern' => 'PHPLIB-438',
+        'aggregate-merge: Aggregate with $merge and available readConcern' => 'PHPLIB-438',
+        'aggregate-out-readConcern: readConcern majority with out stage' => 'PHPLIB-431',
+        'aggregate-out-readConcern: readConcern local with out stage' => 'PHPLIB-431',
+        'aggregate-out-readConcern: readConcern available with out stage' => 'PHPLIB-431',
+        'aggregate-out-readConcern: readConcern linearizable with out stage' => 'PHPLIB-431',
+        'aggregate-out-readConcern: invalid readConcern with out stage' => 'PHPLIB-431',
+        'bulkWrite-arrayFilters: BulkWrite with arrayFilters' => 'Fails due to command assertions',
```

Comment:
Is it possible this only failed because you _weren't_ ignoring command failed/succeeded events? If so, see if my earlier suggested change allows this to pass.

The actual spec test for this PR (`db-aggregate.json`) happens to not include any APM expectations. By coincidence, you're already skipping the other tests since related tickets aren't done. That leaves `bulkWrite-arrayFilters` as the only test that was actually using APM, so it would make sense if the earlier issue was causing this.

Reply:
I've taken the time to step-debug this and the issue is that while we expect a `BSONArray` instance for `arrayFilters` in the bulkWrite update document, we seem to send a `BSONDocument` instead, which causes the constraint failure. I've created https://jira.mongodb.org/browse/PHPLIB-458 to investigate the issue at hand. Along the way, I've also created https://jira.mongodb.org/browse/PHPLIB-459 to update the `DocumentsMatchConstraint` in question to better explained which part of the documents don't match, as I was only able to figure this out by manually checking the objects in question.